cmd/server/pactasrv/analysis.go

```diff
@@ -341,3 +283,88 @@ func (s *Server) RunAnalysis(ctx context.Context, request api.RunAnalysisRequest
+		}
+		return oapierr.Internal("querying analysis for authz failed", zap.Error(err))
+	}
+	as := &authzStatus{
+		primaryTargetID:      string(analysisID),
+		primaryTargetType:    pacta.AuditLogTargetType_Analysis,
+		primaryTargetOwnerID: analysis.Owner.ID,
+		actorInfo:            actorInfo,
+		action:               action,
+	}
+	switch action {
+	case pacta.AuditLogAction_Update, pacta.AuditLogAction_Delete, pacta.AuditLogAction_ReadMetadata:
+		as.isAuthorized, as.authorizedAsActorType = allowIfAdminOrOwner(actorInfo, analysis.Owner.ID)
+	default:
+		return fmt.Errorf("unknown action %q for analysis authz", action)
+
+	}
+	return s.auditLogIfAuthorizedOrFail(ctx, as)
+}
+
+func (s *Server) analysisArtifactDoAuthzAndAuditLog(ctx context.Context, aaID pacta.AnalysisArtifactID, action pacta.AuditLogAction) error {
+	actorInfo, err := s.getActorInfoOrFail(ctx)
+	if err != nil {
+		return err
+	}
+	artifact, err := s.DB.AnalysisArtifact(s.DB.NoTxn(ctx), aaID)
+	if err != nil {
+		if db.IsNotFound(err) {
+			return notFoundOrUnauthorized(actorInfo, action, pacta.AuditLogTargetType_AnalysisArtifact, aaID)
+		}
+		return oapierr.Internal("failed to look up analysis artifact", zap.String("analysis_artifact_id", string(aaID)), zap.Error(err))
+	}
+	aID := artifact.AnalysisID
+	analysis, err := s.DB.Analysis(s.DB.NoTxn(ctx), aID)
+	if err != nil {
+		if db.IsNotFound(err) {
+			return notFoundOrUnauthorized(actorInfo, action, pacta.AuditLogTargetType_AnalysisArtifact, aaID)
+		}
+		return oapierr.Internal("failed to look up analysis for analysis artifact", zap.String("analysis_id", string(aID)), zap.Error(err))
+	}
+
+	as := &authzStatus{
+		primaryTargetID:        string(aID),
+		primaryTargetType:      pacta.AuditLogTargetType_Analysis,
+		primaryTargetOwnerID:   analysis.Owner.ID,
+		secondaryTargetID:      ptr(string(aaID)),
+		secondaryTargetType:    ptr(pacta.AuditLogTargetType_AnalysisArtifact),
+		secondaryTargetOwnerID: ptr(analysis.Owner.ID),
+		actorInfo:              actorInfo,
+		action:                 action,
+	}
+	switch action {
+	case pacta.AuditLogAction_Download:
+		if actorInfo.OwnerID == analysis.Owner.ID {
+			as.isAuthorized, as.authorizedAsActorType = true, ptr(pacta.AuditLogActorType_Owner)
+		} else if artifact.SharedToPublic {
+			as.isAuthorized, as.authorizedAsActorType = true, ptr(pacta.AuditLogActorType_Public)
+		} else if artifact.AdminDebugEnabled {
+			as.isAuthorized, as.authorizedAsActorType = allowIfAdmin(actorInfo)
+		} else {
+			as.isAuthorized, as.authorizedAsActorType = false, nil
+		}
+	case pacta.AuditLogAction_EnableAdminDebug,
+		pacta.AuditLogAction_DisableAdminDebug,
+		pacta.AuditLogAction_EnableSharing,
+		pacta.AuditLogAction_DisableSharing:
+		as.isAuthorized, as.authorizedAsActorType = allowIfOwner(actorInfo, analysis.Owner.ID)
+	case pacta.AuditLogAction_Update,
+		pacta.AuditLogAction_Delete:
+		as.isAuthorized, as.authorizedAsActorType = allowIfAdminOrOwner(actorInfo, analysis.Owner.ID)
```

Comment:
Looks like admins are able to update/delete analysis artifacts regardless of `AdminDebugEnabled`? Same with a few other resources it looks like, is this intended behavior?

Reply:
Yes! Admins can update object metadata, and delete objects, they just cannot **see the actual content of portfolios**, which is why we isolate the checks for ADE to (a) blob => download URL endpoints, and (b) changing of ADE status.